src/core_modules/capture-core-utils/RulesEngine/RulesEngine.js

```diff
@@ -16,80 +17,380 @@ import type {
+            }
+            return { isRuleEffective, rule };
+        })
+        .filter(({ isRuleEffective }) => isRuleEffective)
+        .flatMap(({ rule }) => rule.programRuleActions.map((
+            {
+                data: actionData,
+                programRuleActionType: action,
+                id,
+                location,
+                dataElementId,
+                trackedEntityAttributeId,
+                programStageId,
+                programStageSectionId,
+                optionGroupId,
+                optionId,
+                content,
+            }) => {
+            let ruleEffectData;
+            if (actionData) {
+                const strippedExpression = replaceVariablesWithValues(actionData, variablesHash);
+                const evaluatedRuleEffectData = executeExpression(dhisFunctions, strippedExpression, e => log.warn(`Expression with id rule: action:${id} could not be run. Original condition was: ${rule.condition} - Evaluation ended up as:${strippedExpression} - error message:${e}`));
+                ruleEffectData = trimQuotes(evaluatedRuleEffectData);
+            }
+            return {
+                data: ruleEffectData,
+                id,
+                location,
+                action,
+                dataElementId,
+                trackedEntityAttributeId,
+                programStageId,
+                programStageSectionId,
+                optionGroupId,
+                optionId,
+                content,
+            };
+        }));
+}
+
+
+export default class RulesEngine {
+  variableService: VariableService;
+  dateUtils: DateUtils;
+  processRulesEffects: (
+        effects: ?Array<ProgramRuleEffect>,
+        processType: $Values<typeof processTypes>,
+        dataElements: ?DataElements,
+        trackedEntityAttributes: ?TrackedEntityAttributes
+  ) => ?OutputEffects;
+  generateEffects: (
+    variablesHash: RuleVariables,
+    programRules: ?Array<ProgramRule>,
+    dataElements: ?DataElements,
+    trackedEntityAttributes: ?TrackedEntityAttributes
+  )=> ?OutputEffects;
+
+  constructor() {
+      const valueProcessor = new ValueProcessor(inputValueConverter);
+
+      this.dateUtils = getDateUtils(momentConverter);
+      this.variableService = new VariableService(valueProcessor.processValue);
+      this.processRulesEffects = getRulesEffectsProcessor(
+          convertRuleEffectDataToOutputBaseValue,
+          rulesEffectsValueConverter,
+      );
+
+      this.generateEffects = (variablesHash, programRules, dataElements, trackedEntityAttributes) => {
+          const dhisFunctions = d2Functions(this.dateUtils, this.variableService, variablesHash);
+          const effects = getEffects(dhisFunctions, programRules, dataElements, trackedEntityAttributes, variablesHash);
+          updateVariableHashWhenActionIsAssignValue(effects, variablesHash);
```

Comment:
You have rules depending on the result of other rules (assign action) so updating the variableHash after getting all the effects wouldn't work I think. Maybe I'm missing something though, but this seems suspicious.

Reply:
About this. Lets discuss it. One thing is that I have written tests for this case, and I have also tested it manually after importing the COVID rules but that doesnt mean that it cannot be broken.

Basically if you see [here](https://github.com/dhis2/capture-app/pull/630/files#diff-9c101046a67bb3c0f64cf485a7fa0fddL645) these 3 lines what they do is:
1. the rule is being checked if it is effective.
2. then you iterate though the `programRuleActions` and as a result you will have the `ruleEffects `

While this is fine if you take a closer look in the line [576 of the `executionService`](https://github.com/dhis2/capture-app/blob/cp%2Frefactors-rules-engine%2F09-writes-tests-on-functions/src/core_modules/capture-core-utils/RulesEngine/executionService/executionService.js#L576) you will see that this `if` statement there has a little to do with the getting the effects. What it basically does is updating the `variablesHash` if there is an action called `ASSIGN`.  
In my point of view, the lines from [576](https://github.com/dhis2/capture-app/blob/cp%2Frefactors-rules-engine%2F09-writes-tests-on-functions/src/core_modules/capture-core-utils/RulesEngine/executionService/executionService.js#L576) to [588](https://github.com/dhis2/capture-app/blob/cp%2Frefactors-rules-engine%2F09-writes-tests-on-functions/src/core_modules/capture-core-utils/RulesEngine/executionService/executionService.js#L588) have nothing to do with the function that is being called `getEffects`. 

What I did is I took these lines out of there and made it its own function.

So now instead of having a function that returns some effects _and_  updates the `variablesHash` when action is equal to `ASSIGN`, you now first get all the effects from the function called `getEffects` and then you iterate through the effects[1] and if there is an effect with an action that is equal to `ASSIGN` then you update the `variablesHash`

[1] so yeah now we are iterating though the effects twice instead of once but we have separated the concepts and we made it easier for the reader to understand what is going on and in which order things are taking place. I dont wanna argue about performance here since if you ask me it needs to be millions of effects for this to be a concern.


Does this explains a little bit better why I have done what I have done over there?

Reply:
Don't get me wrong, I like the separation of the concepts.

> So now instead of having a function that returns some effects and updates the variablesHash when action is equal to ASSIGN, you now first get all the effects from the function called getEffects and then you iterate through the effects[1] and if there is an effect with an action that is equal to ASSIGN then you update the variablesHash.

The problem I see is that one program rule in the chain could rely on an updated value in the variableHash from an earlier rule in the chain. Say you have two rules. The first one assigns a value to a variable. In the second rule you're checking the value of this variable in the expression. If the variableHash is updated post execution (after getEffects) the variable value would be wrong when running the second rule.

Reply:
AH!

Reply:
Might be true what you saying. How can we test that scenario. That's the question.